### PR TITLE
[expo-cli] parallelize expo-optimize

### PR DESCRIPTION
**Pull request by mako-taco**
# Why

Addresses #3937

# How

Using `@supercharge/promise-pool`, we're able to run IO-bound tasks in parallel based on the number of available cores.

# Test Plan

I ran `expo-optimize` in my own project. It would be nice to get some fixtures and automated tests for this tool, though.

### CODE REVIEW

**Comment by mako-taco** [comment]
image
things happening in parallel.  some logging info is lost, because the module being used to call `sharp` does not let us specify what to do with stdout.

**Comment by brentvatne** [comment]
hey @mako-taco! thanks for this. it may be worth following up on the sharp api here to ensure we can control the output. you can find the sharp wrapper module here: https://github.com/expo/expo-cli/blob/master/packages/image-utils/src/sharp.ts

**Comment by mako-taco** [comment]
@brentvatne I was mistaken, all the output was from this file.  I've batched it up and it looks better.

image

File: .yarnrc.yml [hunk 1]
```diff
@@ -0,0 +1 @@
+yarnPath: '.yarn/releases/yarn-1.23.0-20210726.1745.cjs'
```
Comment (brentvatne) [inline .yarnrc.yml:1]: can you remove this and the `.yarn/releases` file please?

File: packages/expo-optimize/package.json [hunk 2]
```diff
@@ -36,6 +36,7 @@
     "@expo/config": "6.0.7",
     "@expo/image-utils": "0.3.17",
     "@expo/json-file": "8.2.33",
+    "@supercharge/promise-pool": "^1.8.0",
```
Comment (brentvatne) [inline packages/expo-optimize/package.json:39]: this dependency doesn't look particularly reliable and it's much larger than it needs to be. can you replace this with p-limit instead? https://www.npmjs.com/package/p-limit

**Review by brentvatne — CHANGES_REQUESTED** [review]

File: .yarnrc.yml (hunk 1 above)
Comment (mako-taco) [reply .yarnrc.yml:1]: sorry, `commit -am` got me.  you should consider checking in `.yarn/releases` and an rc file though, it makes development very difficult for those of us who have upgraded yarn to 2+.

File: packages/expo-optimize/src/assets.ts [hunk 3]
```diff
@@ -140,18 +136,31 @@ function filterImages(files: string[], projectRoot: string) {
 }
 
 // Calculate SHA256 Checksum value of a file based on its contents
-function calculateHash(filePath: string): string {
-  const contents = readFileSync(filePath);
+async function calculateHash(filePath: string): Promise<string> {
+  const contents = await new Promise<Buffer>((resolve, reject) =>
```
Comment (EvanBacon) [inline packages/expo-optimize/src/assets.ts:140]: generally we use `await fs.promises.readFile()` since Node.js was dropped earlier this year.

**Review by EvanBacon — APPROVED** [review]
lgtm

**Pull request merged**